File: c/tests/test_companion_session.c
/*
 * test_companion_session.c — the concrete CompanionSession end-to-end: recalls
 * fused memory + user facts into the system prompt, calls the generator, persists
 * the exchange, hands it to the encoder, recalls a prior turn later, reflects the
 * recalled memories in the context. Mirrors the Rust suite
 * companion_session_test.rs (and TS/Go).
 *
 * The capturing generator copies the messages it was handed into a struct the
 * test inspects, and returns a freshly malloc'd canned reply (the session frees
 * it after persisting; the test also gets its own copy back from send()).
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <assert.h>
#include "circle_ai/circle_ai.h"

static char *tdup(const char *s) {
    if (!s) return NULL;
    size_t n = strlen(s) + 1; char *p = (char *)malloc(n); if (p) memcpy(p, s, n); return p;
}

/* Capturing generator state. */
typedef struct {
    const char *reply;
    /* captured on each call */
    char       *last_system;   /* content of msgs[0] */
    char       *last_user;     /* content of msgs[n-1] */
    ca_role_t   first_role;
    size_t      msg_count;
} capturing_gen_t;

static char *capturing_generate(void *user, const ca_chat_message_t *msgs, size_t n) {
    capturing_gen_t *g = (capturing_gen_t *)user;
    free(g->last_system);
    free(g->last_user);
    g->last_system = tdup(n > 0 ? msgs[0].content : "");
    g->last_user = tdup(n > 0 ? msgs[n - 1].content : "");
    g->first_role = n > 0 ? msgs[0].role : CA_ROLE_USER;
    g->msg_count = n;
    return tdup(g->reply); /* session takes ownership */
}

static void gen_reset(capturing_gen_t *g, const char *reply) {
    memset(g, 0, sizeof(*g));
    g->reply = reply;
}
static void gen_free(capturing_gen_t *g) { free(g->last_system); free(g->last_user); }

static ca_episodic_entry_t seed_entry(const char *user_text, const char *assistant_text) {
    ca_episodic_entry_t e;
    memset(&e, 0, sizeof(e));
    e.id = (char *)"seed";
    e.user_text = (char *)user_text;
    e.assistant_text = (char *)assistant_text;
    e.recorded_at_ms = 1735689600000LL;
    return e;
}

static void record_self_fact(ca_self_belief_store_t *beliefs, const char *text) {
    size_t n = 0;
    ca_personal_belief_t *bs = ca_belief_extract(text, "t0", &n);
    for (size_t i = 0; i < n; ++i) ca_self_belief_store_record(beliefs, &bs[i]);
    ca_personal_belief_free_array(bs, n);
}

int main(void) {
    ca_companion_session_options_t opts;
    memset(&opts, 0, sizeof(opts));
    opts.session_id = "s1";
    opts.identity_id = "u1";
    opts.interface_kind = "mobile";
    opts.recall_top_k = 5;

    /* ── injects recalled memories and user facts into the system prompt ── */
    {
        ca_episodic_store_t *ep = ca_episodic_store_create(1000);
        ca_episodic_entry_t e = seed_entry("I have a peanut allergy", "Noted");
        ca_episodic_store_add(ep, &e);
        ca_self_belief_store_t *beliefs = ca_self_belief_store_create();
        record_self_fact(beliefs, "i am vegetarian");

        ca_fused_recall_t *recall = ca_fused_recall_create(
            ca_episodic_store_search_adapter, ep, NULL, NULL, NULL);
        capturing_gen_t g; gen_reset(&g, "Here are some options");
        ca_companion_session_t *s = ca_companion_session_create(
            capturing_generate, &g, ep, recall, NULL, beliefs, &opts);

        char *reply = ca_companion_session_send(s, "what can I eat?");
        assert(reply && strcmp(reply, "Here are some options") == 0);
        free(reply);

        assert(g.first_role == CA_ROLE_SYSTEM);
        assert(strstr(g.last_system, "peanut allergy") != NULL);
        assert(strstr(g.last_system, "vegetarian") != NULL);
        assert(strcmp(g.last_user, "what can I eat?") == 0);

        gen_free(&g);
        ca_companion_session_destroy(s);
        ca_fused_recall_destroy(recall);
        ca_self_belief_store_destroy(beliefs);
        ca_episodic_store_destroy(ep);
    }

    /* ── persists the turn and grows the history ── */
    {
        ca_episodic_store_t *ep = ca_episodic_store_create(1000);
        ca_fused_recall_t *recall = ca_fused_recall_create(
            ca_episodic_store_search_adapter, ep, NULL, NULL, NULL);
        capturing_gen_t g; gen_reset(&g, "ok");
        ca_companion_session_t *s = ca_companion_session_create(
            capturing_generate, &g, ep, recall, NULL, NULL, &opts);

        char *reply = ca_companion_session_send(s, "hello");
        free(reply);
        assert(ca_episodic_store_count(ep) == 1);
        assert(ca_companion_session_history_count(s) == 2);
        assert(strcmp(ca_companion_session_history_role(s, 0), "user") == 0);
        assert(strcmp(ca_companion_session_history_role(s, 1), "assistant") == 0);

        gen_free(&g);
        ca_companion_session_destroy(s);
        ca_fused_recall_destroy(recall);
        ca_episodic_store_destroy(ep);
    }

    /* ── recalls a prior turn on a later turn ── */
    {
        ca_episodic_store_t *ep = ca_episodic_store_create(1000);
        ca_fused_recall_t *recall = ca_fused_recall_create(
            ca_episodic_store_search_adapter, ep, NULL, NULL, NULL);
        capturing_gen_t g; gen_reset(&g, "noted");
        ca_companion_session_t *s = ca_companion_session_create(
            capturing_generate, &g, ep, recall, NULL, NULL, &opts);

        char *r1 = ca_companion_session_send(s, "my favourite colour is blue");
        free(r1);
        char *r2 = ca_companion_session_send(s, "what's my favourite colour?");
        free(r2);

        assert(strstr(g.last_system, "favourite colour is blue") != NULL);

        gen_free(&g);
        ca_companion_session_destroy(s);
        ca_fused_recall_destroy(recall);
        ca_episodic_store_destroy(ep);
    }

    /* ── hands the turn to the background encoder, filling the graph ── */
    {
        ca_episodic_store_t *ep = ca_episodic_store_create(1000);
        ca_knowledge_graph_t *graph = ca_knowledge_graph_create();
        ca_memory_encoder_t *enc = ca_memory_encoder_create(
            ca_kg_extractor_heuristic_adapter, NULL, graph, NULL, NULL, NULL, 0);
        ca_fused_recall_t *recall = ca_fused_recall_create(
            ca_episodic_store_search_adapter, ep, NULL, NULL, NULL);
        capturing_gen_t g; gen_reset(&g, "ok");
        ca_companion_session_t *s = ca_companion_session_create(
            capturing_generate, &g, ep, recall, enc, NULL, &opts);

        char *reply = ca_companion_session_send(s, "remember my dentist appointment");
        free(reply);
        ca_memory_encoder_close(enc);

        size_t n = 0;
        ca_knowledge_triple_t *all = ca_knowledge_graph_all_triples(graph, &n);
        bool found = false;
        for (size_t i = 0; i < n; ++i) if (strcmp(all[i].object, "dentist") == 0) found = true;
        assert(found);
        ca_knowledge_triple_free_array(all, n);

        gen_free(&g);
        ca_companion_session_destroy(s);
        ca_fused_recall_destroy(recall);
        ca_memory_encoder_destroy(enc);
        ca_knowledge_graph_destroy(graph);
        ca_episodic_store_destroy(ep);
    }

    /* ── get_context reflects the memories recalled on the last turn ── */
    {
        ca_episodic_store_t *ep = ca_episodic_store_create(1000);
        ca_episodic_entry_t e = seed_entry("I live in Durban", "Nice");
        ca_episodic_store_add(ep, &e);
        ca_fused_recall_t *recall = ca_fused_recall_create(
            ca_episodic_store_search_adapter, ep, NULL, NULL, NULL);
        capturing_gen_t g; gen_reset(&g, "ok");
        ca_companion_session_t *s = ca_companion_session_create(
            capturing_generate, &g, ep, recall, NULL, NULL, &opts);

        char *reply = ca_companion_session_send(s, "where do I live?");
        free(reply);

        size_t sc = 0;
        const char *const *snips = ca_companion_session_context_snippets(s, &sc);
        bool found = false;
        for (size_t i = 0; i < sc; ++i) if (strcmp(snips[i], "I live in Durban") == 0) found = true;
        assert(found);

        gen_free(&g);
        ca_companion_session_destroy(s);
        ca_fused_recall_destroy(recall);
        ca_episodic_store_destroy(ep);
    }

    printf("All companion session tests passed.\n");
    return 0;
}

File: python/tests/test_companion_session.py
"""test_companion_session.py

Verifies the concrete CompanionSession end-to-end: a turn recalls fused memory +
the user's own facts into the system prompt, calls the generator, persists the
exchange, hands it to the background encoder, recalls it on a later turn, and
streams. Mirrors the TypeScript pilot (companion_session.test.ts) and Go port.
"""
from __future__ import annotations

import uuid
from datetime import datetime, timezone
from typing import AsyncGenerator, Optional

from circle_ai.companion.belief import HeuristicBeliefExtractor, SelfBeliefStore
from circle_ai.companion.companion_types import InterfaceKind
from circle_ai.companion.memory_encoder import CompanionMemoryEncoder
from circle_ai.companion.session import CompanionSession, CompanionSessionOptions
from circle_ai.memory.extractor import HeuristicKnowledgeGraphExtractor
from circle_ai.memory.graph import InMemoryKnowledgeGraph
from circle_ai.memory.episodic_memory import EpisodicMemoryEntry
from circle_ai.memory.in_memory_episodic_store import InMemoryEpisodicStore
from circle_ai.memory.recall import FusedRecall
from circle_ai.models.models import ChatMessage


class CapturingGenerator:
    """Records the prompt it was handed and returns a canned reply / chunks."""

    def __init__(self, reply: str, chunks: Optional[list[str]] = None) -> None:
        self._reply = reply
        self._chunks = chunks
        self.last_messages: list[ChatMessage] = []

    async def generate_async(
        self, messages: list[ChatMessage], options=None
    ) -> str:
        self.last_messages = list(messages)
        return self._reply

    async def stream_async(
        self, messages: list[ChatMessage], options=None
    ) -> AsyncGenerator[str, None]:
        self.last_messages = list(messages)
        for c in self._chunks if self._chunks is not None else [self._reply]:
            yield c


async def _record_self_fact(beliefs: SelfBeliefStore, text: str) -> None:
    bx = HeuristicBeliefExtractor()
    for b in await bx.extract_async(text, "t0"):
        beliefs.record(b)


def _make_session(
    generator,
    episodic: InMemoryEpisodicStore,
    *,
    beliefs: Optional[SelfBeliefStore] = None,
    encoder: Optional[CompanionMemoryEncoder] = None,
) -> CompanionSession:
    recall = FusedRecall(episodic, None)
    return CompanionSession(
        generator,
        episodic,
        recall,
        CompanionSessionOptions(
            session_id="s1",
            identity_id="u1",
            interface=InterfaceKind.MOBILE,
            beliefs=beliefs,
            encoder=encoder,
        ),
    )


def _seed_entry(id: str, user_text: str, assistant_text: str) -> EpisodicMemoryEntry:
    return EpisodicMemoryEntry(
        id=uuid.uuid5(uuid.NAMESPACE_OID, id),
        recorded_at_utc=datetime(2026, 1, 1, tzinfo=timezone.utc),
        user_text=user_text,
        assistant_text=assistant_text,
    )


# ── send path ─────────────────────────────────────────────────────────────────


async def test_injects_recalled_memories_and_user_facts_into_system_prompt() -> None:
    episodic = InMemoryEpisodicStore()
    await episodic.add_async(
        _seed_entry("seed1", "I have a peanut allergy", "Noted")
    )
    beliefs = SelfBeliefStore()
    await _record_self_fact(beliefs, "i am vegetarian")

    gen = CapturingGenerator("Here are some options")
    session = _make_session(gen, episodic, beliefs=beliefs)

    reply = await session.send_async("what can I eat?")
    assert reply == "Here are some options"

    system = gen.last_messages[0]
    assert system.role == "system"
    assert "peanut allergy" in system.content, "recalled memory should be in the prompt"
    assert "vegetarian" in system.content, "user fact should be in the prompt"

    # The user's actual message is the last turn handed to the generator.
    assert gen.last_messages[-1].content == "what can I eat?"


async def test_persists_the_turn_and_grows_the_history() -> None:
    episodic = InMemoryEpisodicStore()
    session = _make_session(CapturingGenerator("ok"), episodic)

    await session.send_async("hello")
    assert await episodic.count_async() == 1
    assert len(session.history) == 2  # user + assistant
    assert session.history[0].role == "user"
    assert session.history[1].role == "assistant"


async def test_recalls_a_prior_turn_on_a_later_turn() -> None:
    episodic = InMemoryEpisodicStore()
    gen = CapturingGenerator("noted")
    session = _make_session(gen, episodic)

    await session.send_async("my favourite colour is blue")
    await session.send_async("what's my favourite colour?")

    system = gen.last_messages[0]
    assert "favourite colour is blue" in system.content, "the earlier turn should be recalled"


async def test_hands_the_turn_to_the_background_encoder_filling_the_graph() -> None:
    episodic = InMemoryEpisodicStore()
    graph = InMemoryKnowledgeGraph()
    encoder = CompanionMemoryEncoder(HeuristicKnowledgeGraphExtractor(), graph)
    session = _make_session(CapturingGenerator("ok"), episodic, encoder=encoder)

    await session.send_async("remember my dentist appointment")
    await encoder.close_async()

    assert any(t.object == "dentist" for t in graph.all_triples()), (
        "the encoder should have extracted the turn into the graph"
    )


# ── stream + context ──────────────────────────────────────────────────────────


async def test_streams_chunks_and_still_persists_the_full_reply() -> None:
    episodic = InMemoryEpisodicStore()
    gen = CapturingGenerator("unused", ["Hel", "lo"])
    session = _make_session(gen, episodic)

    chunks: list[str] = []
    async for c in session.stream_async("hi"):
        chunks.append(c)

    assert chunks == ["Hel", "lo"]
    assert await episodic.count_async() == 1
    assert session.history[1].content == "Hello"  # accumulated reply persisted


async def test_get_context_reflects_memories_recalled_on_last_turn() -> None:
    episodic = InMemoryEpisodicStore()
    await episodic.add_async(_seed_entry("seed1", "I live in Durban", "Nice"))
    session = _make_session(CapturingGenerator("ok"), episodic)

    await session.send_async("where do I live?")
    assert "I live in Durban" in session.get_context().recent_memory_snippets


async def test_agent_async_returns_a_reply_and_persists() -> None:
    episodic = InMemoryEpisodicStore()
    session = _make_session(CapturingGenerator("done"), episodic)
    reply = await session.agent_async("do the thing")
    assert reply == "done"
    assert await episodic.count_async() == 1
